system/lib/wasmfs/wasmfs.cpp

```diff
@@ -6,47 +6,85 @@
 // Current Status: Work in Progress.
 // See https://github.com/emscripten-core/emscripten/issues/15041.
 
+#include "file.h"
+#include "file_table.h"
 #include <emscripten/emscripten.h>
 #include <emscripten/html5.h>
+#include <errno.h>
+#include <mutex>
 #include <stdlib.h>
+#include <utility>
 #include <vector>
 #include <wasi/api.h>
 
 extern "C" {
 
-static std::vector<char> fd_write_stdstream_buffer;
+using namespace wasmfs;
+
+long __syscall_dup2(long oldfd, long newfd) {
+  FileTable::Handle fileTable = FileTable::get();
+
+  auto oldOpenFile = fileTable[oldfd];
+  // If oldfd is not a valid file descriptor, then the call fails,
+  // and newfd is not closed.
+  if (!oldOpenFile) {
+    return -(EBADF);
+  }
+
+  if (newfd < 0) {
+    return -(EBADF);
+  }
+
+  if (oldfd == newfd) {
+    return oldfd;
+  }
+
+  // If the file descriptor newfd was previously open, it will just be overwritten silently.
+  fileTable[newfd] = oldOpenFile;
+  return newfd;
+}
+
+long __syscall_dup(long fd) {
+
+  FileTable::Handle fileTable = FileTable::get();
```

Comment:
`get`ting the handle also acquires the lock for the lifetime of the handle, so that call is doing more than just getting a reference to the singleton file table and would be necessary no matter how the file table was stored. We are also intentionally storing the file table in a static local to get it to be lazily initialized.

Reply:
I see, that makes sense.